### ISSUE TITLE
Problem with mining on private network

### ISSUE THREAD

**Issue by xxxZedxxx** [ethereum/go-ethereum#17555]
Hi!

I have a problem with mining on a private network - got null on miner.start() . 
I use Geth 1.8.15.

```
 C:\>"C:\Program Files\Geth\geth" --dev --rpc --rpcaddr "0.0.0.0" --rpccorsdomain "*" --rpcapi "admin,debug,miner,shh,txpool,personal,eth,net,web3" console
INFO [08-31|00:11:41.685] Maximum peer count                       ETH=25 LES=0 total=25
INFO [08-31|00:11:43.251] Using developer account                  address=0x56174762F8e660C4d106f22c67DC3BfF8d5a0864
INFO [08-31|00:11:43.259] Starting peer-to-peer node               instance=Geth/v1.8.15-stable-89451f7c/windows-amd64/go1.10.3
INFO [08-31|00:11:43.265] Writing custom genesis block
INFO [08-31|00:11:43.269] Persisted trie from memory database      nodes=11 size=1.71kB time=0s gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [08-31|00:11:43.281] Initialised chain configuration          config="{ChainID: 1337 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: <nil> Engine: clique}"
INFO [08-31|00:11:43.295] Initialising Ethereum protocol           versions="[63 62]" network=1337
INFO [08-31|00:11:43.301] Loaded most recent local header          number=0 hash=e2a54b…f26e19 td=1
INFO [08-31|00:11:43.307] Loaded most recent local full block      number=0 hash=e2a54b…f26e19 td=1
INFO [08-31|00:11:43.312] Loaded most recent local fast block      number=0 hash=e2a54b…f26e19 td=1
INFO [08-31|00:11:43.317] Starting P2P networking
INFO [08-31|00:11:43.318] Stored checkpoint snapshot to disk       number=0 hash=e2a54b…f26e19
INFO [08-31|00:11:43.322] RLPx listener up                         self="enode://863836a7120d39574151a976027f057be6176625a7577e396afcf83a9961096b12c6d2adb2dcfbb4b1b9c477af8967796b1167e10dc4df331e39bd48ed1563b5@[::]:55656?discport=0"
INFO [08-31|00:11:43.322] started whisper v.6.0
INFO [08-31|00:11:43.344] IPC endpoint opened                      url=\\\\.\\pipe\\geth.ipc
INFO [08-31|00:11:43.349] HTTP endpoint opened                     url=http://0.0.0.0:8545   cors=* vhosts=localhost
INFO [08-31|00:11:43.356] Transaction pool price threshold updated price=1000000000
INFO [08-31|00:11:43.361] Transaction pool price threshold updated price=1
INFO [08-31|00:11:43.365] Etherbase automatically configured       address=0x56174762F8e660C4d106f22c67DC3BfF8d5a0864
INFO [08-31|00:11:43.372] Commit new mining work                   number=1 sealhash=c4345b…e12271 uncles=0 txs=0 gas=0 fees=0 elapsed=0s
WARN [08-31|00:11:43.372] Block sealing failed                     err="waiting for transactions"
Welcome to the Geth JavaScript console!

instance: Geth/v1.8.15-stable-89451f7c/windows-amd64/go1.10.3
coinbase: 0x56174762f8e660c4d106f22c67dc3bff8d5a0864
at block: 0 (Thu, 01 Jan 1970 01:00:00 CET)
 datadir:
 modules: admin:1.0 clique:1.0 debug:1.0 eth:1.0 miner:1.0 net:1.0 personal:1.0 rpc:1.0 shh:1.0 txpool:1.0 web3:1.0 
```

`personal.newAccount('123')
"0x16ff4185b427fb096db441fd062c58ad4fa35918"`

` miner.start()
null`

**Comment by davux** [comment]
Same problem here, namely the `Commit new mining work` line followed by `Block sealing failed err="waiting for transactions"` on a dev network. Could it be a regression in 1.8.something?

**Comment by rjl493456442** [comment]
@xxxZedxxx @davux 

For the dev mode, geth will choose clique algorithm as default consensus engine and enable mining implicitly. 

For the clique algorithm, it has a configuration parameter called `Period`, which is used to indicate how often to create a block. In the dev mode, the default value for Period is 0, it means that a new block will be created only after the geth node receives the transaction. So it is the reason for `Block sealing failed err="waiting for transactions"`

But you can specify the Period via `dev.Period` CLI flag explicitly by yourselves.

And regarding the `miner.start` returning `null`. The return type is error, so it represents no error happened.

**Comment by davux** [comment]
Sorry, I'm realising now it's different on my side: even after submitting a transaction I still get those logs (i.e. the transaction is not "mined"). So it's unrelated to this specific issue.

**Comment by xxxZedxxx** [comment]
I have added a transaction but still doesn't work
```
INFO [09-03|21:26:47.797] Setting new local account                address=0x5e049AFAdcd54A0951B21A5D487AAAF72360E7D8
INFO [09-03|21:26:47.810] Submitted contract creation              fullhash=0xd9e8a112af296c3b8a6af99489e3c1a4c11f39443f4994b409307b7b9f0faf1c contract=0xA507A8CFe676FA752595111cD362Eef86738599C

> miner.start()
INFO [09-03|21:27:06.535] Transaction pool price threshold updated price=1
nullINF
O > [09-03|21:27:06.542] Commit new mining work                   number=1 sealhash=bdb1cd…4ced6f uncles=0 txs=0 gas=0 fees=0 elapsed=990.5µs
WARN [09-03|21:27:06.542] Block sealing failed                     err="waiting for transactions"
INFO [09-03|21:27:06.560] Commit new mining work                   number=1 sealhash=4c09c1…ba829e uncles=0 txs=1 gas=113087 fees=0 elapsed=18.943ms
INFO [09-03|21:27:06.564] Successfully sealed new block            number=1 sealhash=4c09c1…ba829e hash=43f4c7…d78b22 elapsed=10.969ms
INFO [09-03|21:27:06.582] 🔨 mined potential block                  number=1 hash=43f4c7…d78b22
INFO [09-03|21:27:06.607] Commit new mining work                   number=2 sealhash=18fdd1…aac516 uncles=0 txs=0 gas=0      fees=0 elapsed=24.933ms
WARN [09-03|21:27:06.607] Block sealing failed                     err="waiting for transactions"
```

**Comment by xxxZedxxx** [comment]
I have added the second transaction but still have the same error

```
> personal.newAccount('123')
"0x42aa059a6a3b540a2d2b2f8b08e7a89eeab59efa"
> INFO [09-04|08:46:43.621] Setting new local account                address=0x6a0D2C472Bc8afe1426173c34ecdff606b50670D
INFO [09-04|08:46:43.631] Commit new mining work                   number=1 sealhash=91b710…a95b49 uncles=0 txs=0 gas=0 fees=0 elapsed=0s
WARN [09-04|08:46:43.631] Block sealing failed                     err="waiting for transactions"
INFO [09-04|08:46:43.631] Submitted transaction                    fullhash=0x50822de97ab99973fe38eec8cb20ea1753bb9f6a143a95e98207191e27dd78b2 recipient=0x42Aa059A6a3b540a2d2b2f8b08e7A89eeaB59efA
INFO [09-04|08:46:43.657] Commit new mining work                   number=1 sealhash=aa09f1…4926af uncles=0 txs=1 gas=21000 fees=0 elapsed=26.518ms
INFO [09-04|08:46:43.662] Successfully sealed new block            number=1 sealhash=aa09f1…4926af hash=0b22ad…15a909 elapsed=7.999ms
INFO [09-04|08:46:43.719] 🔨 mined potential block                  number=1 hash=0b22ad…15a909
WARN [09-04|08:46:43.721] Block sealing failed                     err="waiting for transactions"
INFO [09-04|08:46:43.729] Commit new mining work                   number=2 sealhash=c52707…116af7 uncles=0 txs=0 gas=0     fees=0 elapsed=9.032ms

> eth.coinbase
"0x6a0d2c472bc8afe1426173c34ecdff606b50670d"
> miner.start()
null
> miner.stop()
null
> miner.start()
INFO [09-04|08:49:53.433] Transaction pool price threshold updated price=1
nIuNlFlO
[> 09-04|08:49:53.440] Commit new mining work                   number=2 sealhash=8fb7c6…a35b0d uncles=0 txs=0 gas=0     fees=0 elapsed=0s
WARN [09-04|08:49:53.440] Block sealing failed                     err="waiting for transactions"
```
![screenshot_1](https://user-images.githubusercontent.com/31566989/45014712-cd525580-b01f-11e8-9484-893b08a7757e.jpg)

**Comment by vincentserpoul** [comment]
I have the same issue
```
INFO [09-08|10:36:27.685] Loaded local transaction journal         transactions=2 dropped=0
INFO [09-08|10:36:27.685] Regenerated local transaction journal    transactions=2 accounts=1
WARN [09-08|10:36:27.685] Blockchain not empty, fast sync disabled
INFO [09-08|10:36:27.685] Starting P2P networking
INFO [09-08|10:36:27.686] RLPx listener up                         self="enode://1@[::]:38601?discport=0"
INFO [09-08|10:36:27.686] started whisper v.6.0
INFO [09-08|10:36:27.687] IPC endpoint opened                      url=/root/.ethereum/geth.ipc
INFO [09-08|10:36:27.688] HTTP endpoint opened                     url=http://0.0.0.0:9545      cors=moz-extension://36ff4990-4bdd-4d81-bbd7-91fcbcae8d06 vhosts=localhost
INFO [09-08|10:36:27.688] WebSocket endpoint opened                url=ws://[::]:9546
INFO [09-08|10:36:27.688] Transaction pool price threshold updated price=1000000000
INFO [09-08|10:36:27.688] Transaction pool price threshold updated price=1
INFO [09-08|10:36:27.688] Etherbase automatically configured       address=0x1
INFO [09-08|10:36:27.698] Commit new mining work                   number=56 sealhash=7a5c88…3c7758 uncles=0 txs=0 gas=0 fees=0 elapsed=32.281µs
WARN [09-08|10:36:27.698] Block sealing failed                     err="waiting for transactions"
```

and it stays there, not mining my 2 transactions.

```
    image: ethereum/client-go:v1.8.15
    command:
      - "--dev"
      - "--rpc"
      - "--rpcaddr=0.0.0.0"
      - "--rpcport=9545"
      - "--rpcapi=web3,personal,eth,net"
      - "--ws"
      - "--wsaddr=0.0.0.0"
      - "--wsport=9546"
      - "--wsapi=web3,personal,eth,net"
      - "--port=40303"
    ports:
      - "9545:9545"
      - "9546:9546"
      - "40303:40303"
```

**Comment by rjl493456442** [comment]
@vincentserpoul Could you please check the txpool status when your node setup(`tx.status` in console)? My hunch is the 2 transactions are the future transactions which are not ready to be mined.  

And the warning log `Block sealing failed  err="waiting for transactions"`. It is because clique receives an empty block which attempted to be sealed. We have to reject the empty block for instant chain (0 second period), otherwise the chain will grow up explosively. #17614

**Comment by vincentserpoul** [comment]
the 2 tx are indeed in the tx pool, queued
```
pending: 0
queued:2
```

I remember there was the same issue with parity, and their answer was to add "--no-persistent-txqueue" as a param. https://github.com/paritytech/parity-ethereum/issues/7730

The question is then: what makes a tx go in Q and not immediately mined? Is it an issue with the wallet (I'm using metamask)

**Comment by vincentserpoul** [comment]
I just reseted my account in metamask (settings > Reset Account) and as the nonce is then reseted with it, I can now normally submits all new transactions.

**Comment by vincentserpoul** [comment]
@xxxZedxxx which wallet are you using?

**Comment by xxxZedxxx** [comment]
I'm using Mist. I've added a new account but still have the same issue

**Comment by vincentserpoul** [comment]
Can you test with metamask and check if that works? Then you can be sure that the nonce is the issue and not your setup.

**Comment by rjl493456442** [comment]
Fixed in #17614
